src/smefit/loader.py

```diff
@@ -509,36 +512,78 @@ def construct_corrections_matrix(corrections_list, n_data_tot, sorted_keys=None)
 
     Returns
     -------
-        sorted_keys : np.ndarray
-            unique list of operators for which at least one correction is present
         corr_values : np.ndarray
             matrix with correction values (n_data_tot, sorted_keys.size)
     """
 
-    if sorted_keys is None:
-        tmp = [
-            [
-                *c,
-            ]
-            for _, c in corrections_list
-        ]
-        sorted_keys = np.unique([item for sublist in tmp for item in sublist])
     corr_values = np.zeros((n_data_tot, sorted_keys.size))
     cnt = 0
     # loop on experiments
     for n_dat, correction_dict in corrections_list:
         # loop on corrections
         for key, values in correction_dict.items():
-            if "*" in key:
-                op1, op2 = key.split("*")
-                if op2 < op1:
-                    key = f"{op2}*{op1}"
-
             idx = np.where(sorted_keys == key)[0][0]
             corr_values[cnt : cnt + n_dat, idx] = values
         cnt += n_dat
 
-    return sorted_keys, corr_values
+    if rgemat is not None:
+        if len(rgemat.shape) == 3:
+            corr_values = jnp.einsum("ij, ijk -> ik", corr_values, rgemat)
+        else:
+            corr_values = jnp.einsum("ij, jk -> ik", corr_values, rgemat)
+
+    return corr_values
+
+
+def construct_corrections_matrix_quadratic(
+    corrections_list, n_data_tot, sorted_keys, rgemat=None
+):
+    """
+    Construct a unique list of correction name,
+    with corresponding values for quadratic EFT corrections.
+
+    Parameters
+    ----------
+        corrections_list : list(dict)
+            list containing corrections per experiment
+        n_data_tot : int
+            total number of experimental data points
+        sorted_keys: numpy.ndarray
+            list of sorted operator corrections
+
+    Returns
+    -------
+        corr_values : np.ndarray
+            matrix with correction values (n_data_tot, sorted_keys.size, sorted_keys.size)
+    """
+
+    corr_values = np.zeros((n_data_tot, sorted_keys.size, sorted_keys.size))
+    cnt = 0
+
+    # loop on experiments
+    for n_dat, correction_dict in corrections_list:
+        # loop on corrections
+        for key, values in correction_dict.items():
+            op1, op2 = key.split("*")
+            idx1 = np.where(sorted_keys == op1)[0][0]
+            idx2 = np.where(sorted_keys == op2)[0][0]
+
+            # we want to store the values in the upper triangular part of the matrix
+            if idx1 > idx2:
+                idx1, idx2 = idx2, idx1
+
+            corr_values[cnt : cnt + n_dat, idx1, idx2] = values
+        cnt += n_dat
+
+    if rgemat is not None:
+        if len(rgemat.shape) == 3:
+            corr_values = jnp.einsum(
+                "ijk, ijl, ikr -> ilr", corr_values, rgemat, rgemat
+            )
+        else:
+            corr_values = jnp.einsum("ijk, jl, kr -> ilr", corr_values, rgemat, rgemat)
+
+    return corr_values
```

Comment:
Without RG, this is upper triangular indeed. But this property is lost when the RG is included due to the jnp.einsum contraction above. Are we double counting now? Probably not, but I don't see why atm.

Reply:
Mmm, good spot...